src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py

```diff
@@ -220,6 +225,145 @@ def __init__(
+        if self.source_plan is None:
+            return self
+
+        # only select statement can be converted to CTEs
+        if not is_sql_select_statement(self.queries[-1].sql):
+            return self
+
+        def find_duplicate_subtrees(root: SnowflakePlan) -> Set[SnowflakePlan]:
+            """
+            Returns a set containing all duplicate subtrees in query plan tree.
+            The root of a duplicate subtree is defined as a duplicate node, if
+                - it appears more than once in the tree, AND
+                - one of its parent is unique (only appear once) in the tree, OR
+                - it has multiple different parents
+
+            For example,
+                              root
+                             /    \
+                           df5   df6
+                        /   |     |   \
+                      df3  df3   df4  df4
+                       |    |     |    |
+                      df2  df2   df2  df2
+                       |    |     |    |
+                      df1  df1   df1  df1
+
+            df4, df3 and df2 are duplicate subtrees.
+
+            This function is used to only include nodes that should be converted to CTEs.
+            """
+            node_count_map = defaultdict(int)
+            node_parents_map = {root: set()}
+
+            def traverse(node: SnowflakePlan) -> None:
+                node_count_map[node] += 1
+                if node.source_plan and node.source_plan.children:
+                    for child in node.source_plan.children:
+                        if child not in node_parents_map:
+                            node_parents_map[child] = {node}
+                        else:
+                            node_parents_map[child].add(node)
+                        traverse(child)
+
+            def is_duplicate_subtree(node: SnowflakePlan) -> bool:
+                is_duplicate_node = node_count_map[node] > 1
+                if is_duplicate_node:
+                    is_any_parent_unique_node = any(
+                        node_count_map[n] == 1 for n in node_parents_map[node]
+                    )
+                    if is_any_parent_unique_node:
+                        return True
+                    else:
+                        has_multi_parents = len(node_parents_map[node]) > 1
+                        if has_multi_parents:
+                            return True
+                return False
+
+            traverse(root)
+            return {node for node in node_count_map if is_duplicate_subtree(node)}
+
+        # if there is no duplicate node, no optimization will be performed
+        duplicate_plan_set = find_duplicate_subtrees(self)
+        if not duplicate_plan_set:
+            return self
+
+        plan_to_query_map = {}
+        duplicate_plan_to_cte_map = {}
+        duplicate_plan_to_table_name_map = {}
+
+        def build_plan_to_query_map_in_post_order(node: SnowflakePlan) -> None:
```

Comment:
Same here. Unit tests.

Reply:
I found it's difficult/not worthwhile to write unit test for this function because what we're going to test is the generate sql, and I need to create many fake queries in unit tests. It's already tested in integration tests. But I moved them to a util file, which is more clear to review the code